File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/dialogue/adapter/tts/loadbalancer/TtsEndpoint.java
package com.study.webflux.rag.infrastructure.dialogue.adapter.tts.loadbalancer;

import java.time.Duration;
import java.time.Instant;
import java.util.concurrent.atomic.AtomicInteger;

import com.study.webflux.rag.infrastructure.dialogue.adapter.tts.loadbalancer.circuit.EndpointCircuitBreaker;
import com.study.webflux.rag.infrastructure.dialogue.adapter.tts.loadbalancer.circuit.ExponentialBackoffStrategy;

/**
 * TTS 엔드포인트
 */
public class TtsEndpoint {
	private final String id;
	private final String apiKey;
	private final String baseUrl;
	private final AtomicInteger activeRequests;
	private final int maxConcurrentRequests;
	private final Object healthLock = new Object();
	private final EndpointCircuitBreaker circuitBreaker;
	private volatile EndpointHealth health;
	private volatile Instant circuitOpenedAt;
	private volatile double credits = Double.MAX_VALUE;

	public TtsEndpoint(String id, String apiKey, String baseUrl) {
		this(id, apiKey, baseUrl, 10);
	}

	public TtsEndpoint(String id, String apiKey, String baseUrl, int maxConcurrentRequests) {
		this(id, apiKey, baseUrl, maxConcurrentRequests, new ExponentialBackoffStrategy(
			Duration.ofSeconds(5),
			Duration.ofSeconds(300)));
	}

	public TtsEndpoint(String id,
		String apiKey,
		String baseUrl,
		int maxConcurrentRequests,
		ExponentialBackoffStrategy backoffStrategy) {
		this.id = id;
		this.apiKey = apiKey;
		this.baseUrl = baseUrl;
		this.maxConcurrentRequests = maxConcurrentRequests;
		this.health = EndpointHealth.HEALTHY;
		this.activeRequests = new AtomicInteger(0);
		this.circuitOpenedAt = null;
		this.circuitBreaker = new EndpointCircuitBreaker(backoffStrategy);
	}

	public String getId() {
		return id;
	}

	public String getApiKey() {
		return apiKey;
	}

	public String getBaseUrl() {
		return baseUrl;
	}

	public EndpointHealth getHealth() {
		return health;
	}

	public void setHealth(EndpointHealth health) {
		synchronized (healthLock) {
			this.health = health;
			if (health == EndpointHealth.TEMPORARY_FAILURE
				|| health == EndpointHealth.PERMANENT_FAILURE) {
				this.circuitOpenedAt = Instant.now();
			} else if (health == EndpointHealth.HEALTHY) {
				this.circuitOpenedAt = null;
			}
		}
	}

	public int getActiveRequests() {
		return activeRequests.get();
	}

	public int incrementActiveRequests() {
		return activeRequests.incrementAndGet();
	}

	public int decrementActiveRequests() {
		return activeRequests.decrementAndGet();
	}

	public Instant getCircuitOpenedAt() {
		return circuitOpenedAt;
	}

	public EndpointCircuitBreaker getCircuitBreaker() {
		return circuitBreaker;
	}

	public double getCredits() {
		return credits;
	}

	public void updateCredits(double credits) {
		this.credits = credits;
	}

	public boolean isAvailable() {
		return health == EndpointHealth.HEALTHY;
	}

	/**
	 * 요청 수락 가능 여부 확인
	 *
	 * <p>
	 * 다음 조건을 모두 만족해야 요청 수락 가능:
	 * <ul>
	 * <li>엔드포인트 상태가 HEALTHY</li>
	 * <li>동시 요청 수가 최대값 미만</li>
	 * <li>서킷 브레이커가 요청 허용 상태</li>
	 * </ul>
	 *
	 * @return 요청 수락 가능 여부
	 */
	public boolean canAcceptRequest() {
		return health == EndpointHealth.HEALTHY
			&& activeRequests.get() < maxConcurrentRequests
			&& circuitBreaker.allowRequest();
	}

	/**
	 * 엔드포인트 상태
	 */
	public enum EndpointHealth {
		HEALTHY,
		TEMPORARY_FAILURE,
		PERMANENT_FAILURE,
		CLIENT_ERROR
	}

	/**
	 * 실패 유형
	 */
	public enum FailureType {
		TEMPORARY,
		PERMANENT,
		CLIENT_ERROR
	}
}

File: infrastructure/src/main/kotlin/com/miyou/app/infrastructure/dialogue/adapter/tts/loadbalancer/TtsEndpoint.kt
package com.miyou.app.infrastructure.dialogue.adapter.tts.loadbalancer

import com.miyou.app.infrastructure.dialogue.adapter.tts.loadbalancer.circuit.EndpointCircuitBreaker
import com.miyou.app.infrastructure.dialogue.adapter.tts.loadbalancer.circuit.ExponentialBackoffStrategy
import java.time.Duration
import java.time.Instant
import java.util.concurrent.atomic.AtomicInteger

class TtsEndpoint(
    val id: String,
    val apiKey: String,
    val baseUrl: String,
    val maxConcurrentRequests: Int = 10,
    backoffStrategy: ExponentialBackoffStrategy =
        ExponentialBackoffStrategy(
            Duration.ofSeconds(5),
            Duration.ofSeconds(300),
        ),
) {
    private val healthLock = Any()
    private val activeRequestCounter = AtomicInteger(0)
    private val circuitBreakerInternal = EndpointCircuitBreaker(backoffStrategy)

    @Volatile
    private var circuitOpenedAtInternal: Instant? = null
    private var creditsValue: Double = Double.MAX_VALUE

    init {
        require(id.isNotBlank()) { "id must not be blank" }
        require(apiKey.isNotBlank()) { "apiKey must not be blank" }
        require(baseUrl.isNotBlank()) { "baseUrl must not be blank" }
        require(maxConcurrentRequests > 0) { "maxConcurrentRequests must be positive" }
    }

    var health: EndpointHealth = EndpointHealth.HEALTHY
        set(value) {
            synchronized(healthLock) {
                field = value
                when (value) {
                    EndpointHealth.TEMPORARY_FAILURE, EndpointHealth.PERMANENT_FAILURE -> {
                        circuitOpenedAtInternal = Instant.now()
                    }

                    EndpointHealth.HEALTHY -> {
                        circuitOpenedAtInternal = null
                    }

                    EndpointHealth.CLIENT_ERROR -> {}
                }
            }
        }

    constructor(id: String, apiKey: String, baseUrl: String) : this(
        id = id,
        apiKey = apiKey,
        baseUrl = baseUrl,
        maxConcurrentRequests = 10,
        backoffStrategy = ExponentialBackoffStrategy(Duration.ofSeconds(5), Duration.ofSeconds(300)),
    )

    constructor(id: String, apiKey: String, baseUrl: String, maxConcurrentRequests: Int) : this(
        id = id,
        apiKey = apiKey,
        baseUrl = baseUrl,
        maxConcurrentRequests = maxConcurrentRequests,
        backoffStrategy = ExponentialBackoffStrategy(Duration.ofSeconds(5), Duration.ofSeconds(300)),
    )

    val activeRequests: Int
        get() = activeRequestCounter.get()

    val circuitOpenedAt: Instant?
        get() = circuitOpenedAtInternal

    val circuitBreaker: EndpointCircuitBreaker
        get() = circuitBreakerInternal

    val credits: Double
        get() = creditsValue

    fun isHealthy(): Boolean = health == EndpointHealth.HEALTHY

    fun incrementActiveRequests(): Int = activeRequestCounter.incrementAndGet()

    fun decrementActiveRequests(): Int = activeRequestCounter.decrementAndGet()

    fun updateCredits(newCredits: Double) {
        creditsValue = newCredits
    }

    fun canAcceptRequest(): Boolean =
        health == EndpointHealth.HEALTHY &&
            activeRequests < maxConcurrentRequests &&
            circuitBreakerInternal.allowRequest()

    enum class EndpointHealth {
        HEALTHY,
        TEMPORARY_FAILURE,
        PERMANENT_FAILURE,
        CLIENT_ERROR,
    }

    enum class FailureType {
        TEMPORARY,
        PERMANENT,
        CLIENT_ERROR,
    }
}
